### DISCUSSION TITLE
Changes to the CI/CD structure for upcoming 3.13

### DISCUSSION THREAD

**Discussion by col-panic** [elexis/elexis.github.io#1] [Announcements]
Up to now (Aug 15, 2024) all productive building was done in a closed GitLab service sponsored by [Medelexis](https://www.medelexis.ch). In June 24 we decided to do the building of the open-source part within GitHub.

To realize this, the Elexis Organization now hosts an own GitHub runner, dedicated to building the elexis-3-core, elexis-3-base and elexis-server repository. This builder is sponsored again by [Medelexis](https://www.medelexis.ch).

The build is executed within a [build-image](https://github.com/orgs/elexis/packages/container/package/docker-build) which asserts, that always the same build environment is used. A build can
be reliably executed in the same environment as we use at your local computer, provided it runs [Docker](www.docker.com).

The [elexis-server](https://github.com/elexis/elexis-server/pkgs/container/elexis-server) docker image is now also hosted with GitHub.

The storage of artifacts for update sites is done on https://download.elexis.info also sponsored by [Medelexis](https://www.medelexis.ch).

**Comment by rgwch** [comment]
So I tried:

`docker run ghcr.io/elexis/docker-build:2023-09-java17`

which was obviously too simple thought, it ends with:

```
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.040 s
[INFO] Finished at: 2024-08-20T17:47:00+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] No goals have been specified for this build. You must specify a valid lifecycle phase or a goal in the format <plugin-prefix>:<goal> or <plugin-group-id>:<plugin-artifact-id>[:<plugin-version>]:<goal>. Available lifecycle phases are: validate, initialize, generate-sources, process-sources, generate-resources, process-resources, compile, process-classes, generate-test-sources, process-test-sources, generate-test-resources, process-test-resources, test-compile, process-test-classes, test, prepare-package, package, pre-integration-test, integration-test, post-integration-test, verify, install, deploy, pre-clean, clean, post-clean, pre-site, site, post-site, site-deploy. -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/NoGoalSpecifiedException
```
Can you provide a hint how to use the image?
Thanks, Gerry

PS: The obvious solution from the error message doesn't work either:

```
docker run ghcr.io/elexis/docker-build:2023-09-java17 verify
/usr/local/bin/mvn-entrypoint.sh: 50: exec: verify: not found
```

**Reply by debenol1** [reply]
I used:
`docker run --rm --init -v "$(pwd)":/PATH_TO/elexis-3-core -w /PATH_TO/elexis-3-core  ghcr.io/elexis/docker-build:2023-09-java21 xvfb-run mvn -V clean verify -DskipTests -Dmaterialize-products
`

resp.

`docker run --rm --init -v "$(pwd)":/PATH_TO/elexis-3-base -w /PATH_TO/elexis-3-base  ghcr.io/elexis/docker-build:2023-09-java21 -V clean verify -DskipTests`

Whereas PATH_TO points to the directory being checked out from github

**Reply by rgwch** [reply]
This worked, thank you!

**Comment by ngiger** [comment]
Running `xvfb-run mvn -V clean verify  -Dtycho.localArtifacts=ignore`

Fails with

```
Failures: 
  VirtualFileHandle_SmbDirectory_Test.canRead:37
Errors: 
  VirtualFileHandle_Combined_Test.copyFromFileToSmb:59 » Smb Failed to connect: gitlab.medelexis.ch/88.99.132.91
  VirtualFileHandle_SmbDirectory_Test.isDirectory:48 » IO jcifs.smb.SmbException: Failed to connect: gitlab.medelexis.ch/88.99.132.91
  VirtualFileHandle_SmbDirectory_Test.testCopyToAndDelete:122 » Smb Failed to connect: gitlab.medelexis.ch/88.99.132.91
  VirtualFileHandle_SmbDirectory_Test.testCreateAndMoveToAndDelete:70 » Smb Failed to connect: gitlab.medelexis.ch/88.99.132.91
  VirtualFileHandle_SmbDirectory_Test.testListHandles:56 » Smb Failed to connect: gitlab.medelexis.ch/88.99.132.91
  VirtualFileHandle_SmbDirectory_Test.testListHandlesIVirtualFilesystemhandleFilter:146 » Smb Failed to connect: gitlab.medelexis.ch/88.99.132.91
  VirtualFileHandle_SmbDirectory_Test.testMkdirs:154 » Smb Failed to connect: gitlab.medelexis.ch/88.99.132.91
```
Isn't it possible to replace the occurences to gitlab.medelexis.ch? Or open the access?

**Reply by col-panic** [reply]
@ngiger please use the docker image to build Elexis - it ensures, that we all use the same development tools and makes bug reports like this more reproducible. See e.g. https://github.com/elexis/elexis-3-core/blob/858b4a173d716539b04de8a3abc120919ef11b16/.github/workflows/build-p2-site.yaml#L20 for the code used to build with the Github Actions.
